tenseal/cpp/tensors/encrypted_tensor.h

```diff
@@ -0,0 +1,166 @@
+#ifndef TENSEAL_TENSOR_ENCRYPTED_TENSOR_H
+#define TENSEAL_TENSOR_ENCRYPTED_TENSOR_H
+
+#include "tenseal/cpp/context/tensealcontext.h"
+#include "tenseal/cpp/tensors/utils/matrix_ops.h"
+#include "tenseal/cpp/tensors/utils/utils.h"
+#include "tenseal/cpp/utils/proto.h"
+#include "tenseal/cpp/utils/serialization.h"
+
+namespace tenseal {
+
+using namespace seal;
+using namespace std;
+
+/**
+ * EncryptedTensor<plain_t, plain_container_t, encrypted_t> interface - A
+ *generic API for encrypted tensor operations.
+ * @param plain_t: root plaintext datatype for representing data(double, int64
+ *etc).
+ * @param plain_container_t: complete plaintext representation(for example,
+ *vector<double>).
```

Comment:
I can add a PlainTensor class that wraps the plain_container_t, and will have constructors for any data type we want to support. 

It would help me to know what abstract operations should I add to the Encrypted/PlainTensor classes.

So far, there should be:
 - shape/reshape?
 - broadcast?

should I add tensorflow operations to the interface, as a reference?

Reply:
If we add a single class, how could we distinguish between vectors and multi-dimensional tensors? using the shape only? So operations like add_plain would need to check the shape and see if they can operate with that specific tensor.

For the operations, I guess we can start with just a small subset that is required now, and add as we need them.

Reply:
So, in almost every implementation, tensors have the following properties:
 - **an 1-D vector** for internal representation, of size D * W * H
 - **a size/shape tuple**, with the current view of the 1-D vector, for 3D let's say (D, H, W)
 - **a stride stuple** meaning how much you jump for every dimension in the 1D representation, in the previous example (H * W, W, 1).
 - **the data type**, that plain_t, int64_t/double in our cases.
 - **layout of the data**, if we want to use a map instead of vector, no idea.
 - **device**, if we support CPU/CUDA etc.

right? 
we do not store the data reshaped in memory, we use those strides and shape to jump in the 1D representation.
This is taken from the PyTorch internals http://blog.ezyang.com/2019/05/pytorch-internals/


So, I should add a plaintext tensor with those properties.
And the encrypted one should have similar properties, to handle the strides/shapes/views.

In the current implementation, the encrypted tensor now has a broadcast operation.

Does this make sense?

Reply:
Yeah absolutely, this is really great, and that's what I was also planning to implement for the arbitrary dimension tensor. It seems like all tensor library are using this representation, and it's way more efficient when we need to reshape, as only those steps need to be updated instead of moving data around.

Reply:
I added a PlainTensor class to handle the vector/matrix parameters for TenSEAL.

It's nice that we can easily make it compatible with PyTorch/Tensorflow tensors as well.
Let me know what you think, please.